server.js

```diff
@@ -77,6 +114,108 @@ app.get('/notes', (req, res) => {
   res.json(Object.values(getNotes()));
 });
 
+//Get the tags for a note
+app.get('/notes/tags/:id', (req,res) => {
+  const note = getNote(req.params.id);
+  
+  //404 if a note doesn't exist
+  if (!note)
+    return res.sendStatus(404);
+
+  //return an empty array if tags hasn't been defined
+  res.json(note.tags || []);
+});
+
+app.post('/notes/tags/:id', (req,res) => {
+  const note = getNote(req.params.id);
+
+  //404 if a note doesn't exist
+  if (!note)
+    return res.sendStatus(404);
+
+  //Get the tag array from the request
+  const tagsFromQueryBody = req.body;
+
+  //If the body of the request isn't an array return a 400
+  if (!Array.isArray(tagsFromQueryBody))
+    return res.status(400).send("you must pass in an array of tags");
+
+  //Ensure the note the tags field and it's an array
+  if (!Array.isArray(note.tags))
+    note.tags = [];
+
+  tagsFromQueryBody.map(tag => {
+    //If the incoming item isn't a string, then it's not a valid tag
+    if (typeof tag !== "string") return;
+    
+    //If the note already has this tag, then we don't need to add it
+    //this will also dedup user entered tags
+    if (note.tags.includes(tag)) return;
+
+    note.tags.push(tag);
+  });
+
+  res.json(note);
+});
+
+app.delete('/notes/tags/:id', (req, res) => {
+  const note = getNote(req.params.id);
+
+  //404 if a note doesn't exist
+  if (!note)
+    return res.status(404);
+
+  //Ensure the note the tags field and it's an array
+  if (!Array.isArray(note.tags))
+    note.tags = [];
+
+  //If the body of the request isn't an array return a 400
+  const tagsToRemove = req.body;
+  if (!Array.isArray(tagsToRemove))
+    return res.status(400).send("you must pass in an array of tags");
+
+  //Remove the requested tags from the array
+  note.tags = note.tags.filter(tag => !tagsToRemove.includes(tag));
+
+  res.json(note);
+});
+
+//Helper method to compare two strings, also case insensitive
+const insensitiveCompare = (a, b) => {
+  return new RegExp(`^${a}$`,"i").test(b);
+};
+
+//Supports ?q=searchtext
+app.get('/notes/search', (req, res) => {
+  const notes = Object.values(getNotes());
+
+  //Get the search text from the query string
+  const searchText = req.query.q;
+
+  //If they're not looking for anything,
+  //we don't need to send them anything
+  if (searchText) {
+    //In this simple search we'll check through the tags
+    //to see if any of them match the supplied query
+    const results = notes.filter(note => {
+      //This variable is for debugability
+      let isMatch = Array.isArray(note.tags) 
+        && note.tags.filter(tag => insensitiveCompare(tag, searchText)).length > 0
+
+      return isMatch;
+    });
+
+    //If we have matches then we're done!
+    if (results && results.length)
+      return res.json(results);
+  }
+    
+  //If they're here then they either didn't
+  //supply a search term, or there were no
+  //matching records
+  res.status(404).json([]);
+});
```

Comment:
The order in which the routes are defined makes a difference. I had added the search route at the end which is why I was running into the null ref exception. It was getting picked up by the save note route and using "search" as the id. This won't match with a note and there's no null check.